uamqp/error.py

```diff
@@ -182,7 +319,7 @@ def __init__(self, condition, description=None, info=None):
         super(AMQPConnectionRedirect, self).__init__(condition, description=description, info=info)
 
 
-class AMQPSessionError(AMQPException):
+class AMQPSessionError(AMQPConnectionError):
```

Comment:
I'm not sure this and the LinkDetach should actually be ConnectionErrors?
I feel like there's a terminology overload here... I thought 'Connection' Error might relate to the Connection layer, but here we're implying that it's related to the network connection?

Reply:
aha, that's true, multiple "connection" concepts here:

1. the general connection, whenever a client fails to communicate with the service due to amqp link/session/connection or tcp connection error, it's called a connection error.
2. the tcp connection
3. the amqp connection

we're using the same AMQPConnectionError for three different connection concepts..

I do prefer separating them out.

maybe something like the following:

class ConnectionError
class TransportConnectionError(ConnectionError)
class AMQPConnectionError(ConnectionError)
class AMQPSessionError(ConnectionError)
class AMQPLinkError(ConnectionError)

Reply:
This seems a bit much....
Question - if there's an underlying connectivity error/tcp connection failure - what's going to fail first? Will the Session fail first, followed by the AMQP Connection layer? Or will the AMQP Connection layer fail, which in turn will bring down all it's Sessions and Links?

I would probably just start with two: ConnectionError and SessionError.
A ConnectionError indicates the termination of the AMQP Connection, a SessionError indicates the termination of the Session. Both would simply inherit from the base AMQPError.

All of our errors will have a Code attribute to indicate the underlying reason, so we might have something like `ConnectionError.code == SocketFailure` or something.

Reply:
I like your proposal which is simpler and clearer, changed current error hierarchy.